README.md
# orbitez

P2E canvas game with NFT pass/skin and DeFi mechanics 
1. **Main page**:
  - Staging - https://staging.orbitez.io
  - Production - https://orbitez.io

2. **Dashboard**:
  - Staging - https://staging.orbitez.io/dashboard
  - Production - https://orbitez.io/dashboard

## Tech stack

- Back-end: Next.js
- Front-end: React + Canvas
- Smart-contracts: Archetype
- Testing: jest, completium-cli
- Deploy: Docker, letsencrypt

Smart-contract address (old one):
~~https://better-call.dev/ghostnet/KT1QQnpFLKtUwLDTPofnudfPdmCuBmtmQkrz/operations~~

Smart-contract address (actual, connected to orbitez.io):
https://better-call.dev/ghostnet/KT1WRAucboJ8yD7fFtEpS7vJ1bVs9qaLvm8Q/operations

## Game Universe

The player will have to take part in the process of forming solar systems. They will fight for survival by controlling a small little planet, which is gradually gaining mass by collecting particles of matter and other small celestial bodies.
The speed in the game is measured not in minutes, but in hundreds of thousands of years. Thus, one match lasts about 100 thousand years.
Time in the game universe runs its course. Inevitably there appears technologically advanced civilizations. Over time, the game will change Epoches. The initial set of planets represents the First Epoch of "natural planets". The Second Epoch of "wandering planets" will contain worlds modified by a technically advanced civilization. On them, you will find Space elevators, planetary engines, Bussard ramjets, etc. The Third Epoch of "artificial planets" will consist of fully artificial cosmic bodies, such as Dyson spheres and Ringworlds.

## Transactions flow | Mechanics

- Mint or buy a unique orbitoid for 1tez to get into the game (you can play without owning a planet, your game skin will be the default planet).
- Enter battleground lobby with orbitoid NFT and choose bet type: 1tez (production room size = 3 players, staging room size = 10 players with 7 bots. First 4 bots join after 30 second each, then 1 bot after 1 minute + 1 bot after 1.5 minutes + last bot after 2 minutes).
- 77% of the battleground bank goes to the TOP 3 players at the time of baking the final block (1 round = 15 blocks in Tezos chain, but it's customizable).
- 3% goes to first player who called "Claim all".
- 10% goes to admin address.
- 10% goes to the owner of the server on which the game took place (use our [1-click game server deployment feature](#game-server-deployments) to have your own server!).

![bounty_explain](https://user-images.githubusercontent.com/4786779/221579327-1c8abb2b-b75f-4c79-8f22-e8448f478d8f.png)

## Game Server Deployments
Orbitez allows you to deploy your own game server and optionally a Tezos Node in one click on a cloud Provider of your choice (Currently fully supported are: DigitalOcean. In progress: AWS, GCP, Azure).

Once the server is created you can configure the server parameters and register it in the smart contract of the game - this will allow you to receive 50% of all the fees we collect on every game played.

We leverage Docker hub to securely store images of the orbitez.io game server. 

## Tezos node deployment
Mainnet nodes are often overloaded and/or have high latencies and with many popular projects it gets nearly impossible to mint anything unless you're leveraging your own node.
Our goal is to make deployment of your own Tezos node as simple as possible for the user. Tick a checkbox when deploying a game server with us and we'll bring up your very own Tezos node you can plug & play directly into your Temple wallet. All HTTPS certs are taken care of by us: your own node will have a dedicated HTTPS endpoint based on the orbitez.io subdomain.

## Flex points / FAQs
1. How do we prevent forgery of the leaderboard?

We run a custom Oracle on the back-end server which communicates with the game server and signs the leaderboard object. Front-end only receives signed and packed encoded values of the game state when the game ends - this ensures consistency of data end to end.

2. How do we handle flakiness and slowness of IPFS gateways in global setup?

IPFS gateways have proven to be quite inconsistent and unreliable. Depending on where you are and how loaded gateways are some may perform significantly better than others.
We selected the top 10 IPFS gateway providers and perform a race-test. Whichever gateway provider wins the `Promise.race` gets to serve us our beautiful generative planets inside of the game.


## Planned

- [x] Animated Orbitoid(planet) in action!
- [ ] Play with your own FA2 NFT as a skin in orbitez game (hold 100 ORB token to unlock this feature)
- [ ] ORB token farming (incentivize farmers to bet in LP as good as players to farm with ORB rewards)
- [ ] Liquidity Baking LP(SIRS) bets
- [ ] Proof of leaderboard (Merkle tree -> own tzstamp server). 
- [ ] Each block send tx with merkle tree of all player key events in the game
- [ ] Federative servers(matrix protocol or simmilar) with state replication 

## Gameplay

Fight with players from all over the world as you try to become the largest Planet in a solar system! Control your tiny planet and eat other players to grow larger. Mint your own, unique generative planet as NFT to enter the battlefield!


## 💻 Quick start

### 1) Clone the repository

```
git clone https://github.com/stepandra/orb && cd orb
```

### 2) Requirements:

- [node](https://nodejs.org) installed, version 14 or later.

### 3) Build:

```
cd ./orbitez && npm i &&
npm run build
```

### 4) Start the app:

```
npm start -- -p 3000
```

The app will be available at http://localhost:3000

Note: some functions will not be available in your local app instance since they require secret env variables:
- bot addition feature;
- 1-click game server deployment feature;
- leaderbord signing at the end of the game.

### 4) (Optional) Run for development

- Start the dev server (Next.js):
  ```
  cd ./orbitez && npm i &&
  npm run dev
  ```

  The app will be available at http://localhost:3000
  
- To start a local instance of a game server:
  1. Enter the following commands:
     ```
     cd ./server && yarn &&
     STAGE="local" node src/index.js
     ```
  2. Create an `.env` file in the `orbitez` directory with the following contents:
     ```
     NEXT_PUBLIC_STAGE=local
     ```
     
  By clicking the "PLAY" button on the `/dashboard` page, you will be able to join your own local dev game server.
  

- Or you can boot up a local instance of the latest release of the orbitez server and connect it to the room in contract you own:
  ```
  docker run --env CONTRACT_ADDRESS=KT1QQnpFLKtUwLDTPofnudfPdmCuBmtmQkrz --env SERVER_NAME=${roomName} -d -p 8080:8080 -p 88:88 orbitez/orb-game-server-main:latest
  ```

### 5) Contract tests
#### Requirements:
Install tezos-client(MacOS):
```
brew tap serokell/tezos-packaging-stable https://github.com/serokell/tezos-packaging-stable.git
brew install tezos-client
```

Install tezos-client(Ubuntu):
```
sudo add-apt-repository ppa:serokell/tezos && sudo apt-get update
sudo apt install tezos-client -y
```

Install completium-cli and init it in root directory:
```
npm i @completium/completium-cli -g
completium-cli init
``` 

Install node packages in root directory:
```
npm i
```
Create .env file with `ORACLE_PRIVATE_KEY` and `ORACLE_PUBLIC_KEY` variables:
```
touch .env && echo "ORACLE_PRIVATE_KEY=''" >> .env && echo "ORACLE_PUBLIC_KEY=''" >> .env
```
Run test
```
npm run test-contracts
```
<img width="649" alt="Снимок экрана 2023-01-03 в 23 32 55" src="https://user-images.githubusercontent.com/4786779/210452341-35f8e86b-19f6-4cea-8680-d6a400390998.png">


## Contract entrypoints
- `clear_all` -  clear all assets except `leaderboard` [admin]
- `destroy_server` - remove selected server from storage [admin]
- `remove_room` - remove selected room on server [server_manager]
  - Parameters
    - room_idn string
    - server_idn string
- `create_server`
  - Parameters
    - serverd string
    - manag address
    - room_idx string
    - serverurl string
    - bet_size mutez
    - size_v nat
    - game_duration_v nat
- `end_game`
  - Parameters
    - parameter pair
    - room_idb string
    - serverid string
    - packed_outcome bytes
    - signed_outcome signature
- `enter_room`
  - Parameters
    - room_idv string
    - serverid string
- `refund`
  - Parameters
    - room_idq string
    - server_id string


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

_No changes since the latest release._

## [0.3.3] - 2023-03-08

### Changed

- Refactored planet rendering:
  - Implemented the ability to render multiple different planets at the same time;
  - Added the ability to render directly into a custom canvas (via passing of the canvas element ref);
  - Added the ability to render by-frame into internal (not mounted) canvas with specified parameters and returning the canvas for custom drawing;
  - Improved `<Planet />` component, which now can be used without any preliminary setup, by simply providing a `mintHash` prop;
  - Improved `<PlanetDataList />` component;
  - Redesigned planet rendering initialization and getting of planet features to be fully synchronous instead of asynchronous;
  - Added the ability to render a planet without a background and take 100% of the width and height of the canvas;
- Game optimizations to ensure stable high FPS:
  - Implemented usage of shared single WebGl context to improve performance;
  - Implemeted reusing of a single planet render for all smaller planet cells;
  - Implemented limiting amount of food cells to be rendered on a large zoom out;
  - Limited max zoom-out;
  - Optimized food skin rendering;
  - Implemented limiting of planet rendering FPS animation;
  - Disabled jelly physics;
- Server config:
  - Max zoom out changed from 0.15 to 0.35;
  - Min food size changed from 5 to 1;
  - Max food size changed from 15 to 25;
  - Food amount reduced from 15000 to 3000;
  - Eating food gives 3 times more of foods original mass;
  - Amount of virus cells changed from 50 to 30;
  - Max size of player increased from 400 to 489.897948557;
  - Max cells of player reduced from 16 to 10;
  - Player speed increased by 1.5 times (1 -> 1.5).

### Added

- Animated planets in-game.

### Removed

- jQuery dependency;
- seedRandom dependency;
- Loading of skins via IPFS.

### Fixed

- Bots requests retries sometimes failing bug;
- Game progress timer optimistic updates exceeding the limits bug.

## [0.3.2] - 2023-02-28

### Added

- Implemented the ability to join local dev game server via usage of env variables;
- New non-linear bot addition logic.

### Fixed

- Game progress timer receiving not valid data bug.

## [0.3.1] - 2023-02-27

### Changed

- Updated contract address to: `KT1QQnpFLKtUwLDTPofnudfPdmCuBmtmQkrz`;
- Updated README.md.

## [0.3.0] - 2023-02-16

### Changed

- Improved `<WaitingRoom />` page component by implementing triggering of re-render only when the player list changes, instead of re-rendering every time the data is fetched;
- Improved killing of not allowed players on the game server (checking and killing of unauthorized players are now carried out on every game tick);
- Modified Orbitez Docker Image CI workflow for a manual trigger only;
- Updated launch scripts;
- Updated `README.md` (`staging.orbitez.io` -> `orbitez.io`);
- Reduced the playing area on the game server by half (from `200000000` to `100000000`);
- Updated the old contract address to the new one.

### Added

- **Bot addition functionality**, if players are waiting too long:
  - Created 7 bots and supplied them with `100 XTZ` each from the faucet and a planet NFT;
  - Created `/api/add-bot/{serverName}` POST API endpoint for requesting bot addition;
  - Implemented bot addition logic;
  - Implemented caching of pending bot addition requests to prevent double requests;
  - Implemented storage of bots' private keys and addresses in env variables.
  - Implemented bot requesting logic in `<WaitingRoom />` page component:
    - request delay will be calculated based on latest player joining time;
    - added request retrying;
  - Implemented bot spawning / killing logic on the game server;
  - Implemented the display of bot skins;
- Set up unit testing with `Jest` and `React Testing Library`;
- Added some unit tests;
- Created a production workflow for deployment to production, which can only be triggered manually from the `Actions` tab;
- Implemented the possibility to launch game and servers on custom ports;
- Implemented **Server Selector**;
- Added `@mui/icons-material` package (for icons).

### Fixed

- Fixed using of `http` over `https` with `statsUrl` in `/api/get-signed-leaderboard`;
- Fixed leaderboard not being sorted on `/last-game-stats page`;
- Fixed not returning a value in `useContractServers.js` file `contractServerList.findIndex()` array method;
- Fixed local git repos on deployment machine being out of sync with origin;
- Added auth to git fetch in workflows;
- Fixed overflowing grid areas in `<Dashboard />` page component.

## [0.2.2] - 2023-02-28

### Added

- Implemented the ability to join local dev game server via usage of env variables.

### Fixed

- Game progress timer receiving not valid data bug.

## [0.2.1] - 2023-02-27

### Changed

- Updated contract address to: `KT1QQnpFLKtUwLDTPofnudfPdmCuBmtmQkrz`;
- Updated README.md.

## [0.2.0] - 2023-02-12

### Changed

- Updated contract address to: `KT1WEVRHFcRq8n9iefMNm2s9P2gbB1d5gVzH`;
- Optimized and cleaned-up `useTezos()` hook and `<Dashboard />` page component;
- Improved "connecting" modal layout in `/hud` page;
- Removed game server config from `.gitignore`;
- Implemented `<WaitingRoom />` and `<Hud />` components optimizations.

### Added

- Implemented automatic setting up of subdomains on orbitez.io and Let's Encrypt SSL certificates for user deployed game servers;
- Kicking of duplicated and unregistered users from the game server;
- Implemented restricted access to the `/hud` route. Access conditions:
  - being logged in;
  - having a selected server;
- Storing the selected server in the app context and localStorage (instantly synced);

server/src/config.js
module.exports = Object.seal({
  // MultiOgarII configurations file
  // Lines starting with // are comments

  // [NOTES]
  // MultiOgarII uses cell size instead of cell mass to improve performance!
  // In order to get the cell size from mass value, you need to calculate using this formula:
  //    size: SQRT( mass * 100 )
  //
  // For example, to set start mass: 43:
  //     size: SQRT( 43 * 100 ): SQRT( 4300 ): 65.57
  // Set playerStartSize: 66
  //
  // Also, you can use the following syntax to specify mass:
  //     playerStartSize: massToSize(43)
  // It will be automatically converted to 66

  // [LOGGING]
  // logVerbosity: Console log level (0=NONE// 1=FATAL// 2=ERROR// 3=WARN// 4=INFO// 5=DEBUG)
  // logFileVerbosity: File log level
  "logVerbosity": 4,
  "logFileVerbosity": 5,

  // [SERVER]
  // serverTimeout: Seconds to keep connection alive for non-responding client
  // serverMaxConnections: Maximum number of connections to the server. (0 for no limit)
  // serverPort: Server port which will be used to listen for incoming connections
  // serverBind: Server network interface which will be used to listen for incoming connections (0.0.0.0 for all IPv4 interfaces)
  // serverTracker: Set to 1 if you want to show your server on the tracker http://ogar.mivabe.nl/master (check that your server port is opened for external connections first!)
  // serverGamemode: Gamemodes: 0: FFA, 1: Teams, 2: Experimental, 3: Rainbow
  // serverBots: Number of player bots to spawn (Experimental)
  // serverViewBase: Base view distance of players. Warning: high values may cause lag! Min value is 1920x1080
  // serverMinScale: Minimum viewbox scale for player (low value leads to lags due to large visible area for big cell)
  // serverSpectatorScale: Scale (field of view) used for free roam spectators (low value leads to lag, vanilla: 0.4, old vanilla: 0.25)
  // serverStatsPort: Port for the stats server. Having a negative number will disable the stats server.
  // serverStatsUpdate: Update interval of server stats in seconds
  // mobilePhysics: Whether or not the server uses mobile agar.io physics
  // badWordFilter: Toggle whether you enable bad word filter (set to 0 to disable)
  // serverRestart: Toggle whether you want your server to auto-restart in minutes. (set to 0 to disable)
  "serverTimeout": 300,
  "serverMaxConnections": 500,
  "serverPort": process.env.SERVER_PORT || 8080,
  "serverBind": "0.0.0.0",
  "serverTracker": 0,
  "serverGamemode": 0,
  "serverBots": 0,
  "serverViewBaseX": 1920,
  "serverViewBaseY": 1080,
  "serverMinScale": 0.35,
  "serverSpectatorScale": 0.4,
  "serverStatsPort": process.env.STATS_PORT || 88,
  "serverStatsUpdate": 60,
  "mobilePhysics": 0,
  "badWordFilter": 1,
  "serverRestart": 0,

  // [CLIENT]
  // serverMaxLB: Controls the maximum players displayed on the leaderboard.
  // serverChat: Allows the usage of server chat. 0: no chat, 1: use chat.
  // serverChatAscii: Set to 1 to disable non-ANSI letters in the chat (english only)
  // serverName: Server name, for example "My great server"
  // serverWelcome1: First server welcome message
  // serverWelcome2: Second server welcome message (optional, for info, etc)
  // clientBind: Only allow connections to the server from specified client (eg: http://agar.io - http://mywebsite.com - http://more.com) [Use ' - ' to seperate different websites]
  "serverMaxLB": 500,
  "serverChat": 0,
  "serverChatAscii": 1,
  "separateChatForTeams": 0,
  "serverName": "Orbitez.io",
  "serverWelcome1": "Welcome to Orbitez.io!",
  "serverWelcome2": "",
  "clientBind": "",

  // [ANTI-BOT]
  // serverIpLimit: Controls the maximum number of connections from the same IP (0 for no limit)
  // serverMinionIgnoreTime: minion detection disable time on server startup [seconds]
  // serverMinionThreshold: max connections within serverMinionInterval time period, which will not be marked as minion
  // serverMinionInterval: minion detection interval [milliseconds]
  // serverScrambleLevel: Toggles scrambling of coordinates. 0: No scrambling, 1: lightweight scrambling. 2: full scrambling (also known as scramble minimap), 3 - high level scrambling (no border)
  // playerBotGrow: Cells greater than 625 mass cannot grow from players under 17 mass (set to 1 to disable)
  "serverIpLimit": 4,
  "serverMinionIgnoreTime": 30,
  "serverMinionThreshold": 10,
  "serverMinionInterval": 1000,
  "serverScrambleLevel": 0,
  "playerBotGrow": 0,

  // [BORDER]
  // Border size (vanilla 14142.135623730952)
  "borderWidth": 10000,
  "borderHeight": 10000,

  // [FOOD]
  // foodMinSize: vanilla 10 (mass: 10*10/100: 1 mass)
  // foodMaxSize: vanilla 20 (mass: 20*20/100: 4 mass)
  // foodAmount: The number of food to spawn
  // foodMassGrow: Enable food mass grow ?
  // foodMassEatingCoef: Food eating mass multiplier - how much mass cells gain...
  // .. from eating food (1: default, 2: two times more mass)
  "foodMinSize": 1,
  "foodMaxSize": 25,
  "foodAmount": 3000,
  "foodMassGrow": 1,
  "foodMassEatingCoef": 3,

  // [VIRUSES]
  // virusMinSize: Minimum virus size. (vanilla: mass: val*val/100: 100 mass)
  // virusMaxSize: Maximum virus size (vanilla: mass: val*val/100: 200 mass)
  // virusMaxPoppedSize: Maximum size a popped cell can have
  // virusEqualPopSize: Whether popped cells have equal size or not (1 to enable)
  // virusAmount: Amount of viruses to spawn
  // virusMaxAmount": Maximum amount of viruses allowed ingame.
  // motherCellMaxMass: Maximum amount of mass a mothercell is allowed to have (0 for no limit)
  // virusVelocity: Velocity of moving viruses (speed and distance)
  // virusMaxCells: Maximum cells a player is allowed to have from virus splits (0 for playerMaxCells)
  "virusMinSize": 100,
  "virusMaxSize": 100,
  "virusMaxPoppedSize": 60,
  "virusEqualPopSize": 0,
  "virusAmount": 30,
  "virusMaxAmount": 100,
  "motherCellMaxMass": 0,
  "virusVelocity": 780,
  "virusMaxCells": 0,

  // [EJECTED MASS]
  // ejectSize: vanilla: mass: val*val/100: 13 mass?
  // ejectSizeLoss: Eject size which will be substracted from player cell (vanilla: mass: val*val/100: 18 mass?)
  // ejectCooldown: Tick count until a player can eject mass again in ticks (1 tick: 40 ms)
  // ejectSpawnPercent: Chance for a player to spawn from ejected mass. 0.5: 50% (set to 0 to disable)
  // ejectVirus: Whether or not players can eject viruses instead of mass
  // ejectVelocity: Velocity of ejecting cells (speed and distance)
  "ejectSize": 36.06,
  "ejectSizeLoss": 42.43,
  "ejectCooldown": 3,
  "ejectSpawnPercent": 0.5,
  "ejectVirus": 0,
  "ejectVelocity": 780,

  // [PLAYERS]
  // Reminder: MultiOgarII uses cell size instead of mass!
  //       playerStartMass replaced with playerStartSize
  //
  // playerMinSize: Minimum size a player cell can decay too. (vanilla: val*val/100: 10 mass)
  // playerMaxSize: Maximum size a player cell can have before auto-splitting. (vanilla: mass: val*val/100: 22500 mass)
  // playerMinSplitSize: Mimimum size a player cell has to be to split. (vanilla: mass: val*val/100: 35 mass)
  // playerMinEjectSize: Minimum size a player cell has to be to eject mass. (vanilla: mass: val*val/100: 35 mass)
  // playerStartSize: Start size of the player cell. (vanilla: mass: val*val/100: 10 mass)
  // playerMaxCells: Maximum cells a player is allowed to have. (vanilla is 16)
  // playerSpeed: Player speed multiplier (1: normal speed, 2: twice the normal speed)
  // playerRecombineTime: Base time in seconds before a cell is allowed to recombine (vanilla: 30 seconds)
  // playerDecayRate: Amount of player cell size lost per second
  // playerDecayCap: Maximum mass a cell can have before it's decayrate multiplies by 10. (0 to disable)
  // playerDisconnectTime: Time in seconds before a disconnected player's cell is removed from the server (Set to -1 to never remove)
  // splitVelocity: Velocity of splitting playercells (speed and distance)
  "playerMinSize": 31.6227766017,
  "playerMaxSize": 489.897948557,
  "playerMinSplitSize": 59.16079783,
  "playerMinEjectSize": 59.16079783,
  "playerStartSize": 31.6227766017,
  "playerMaxCells": 10,
  "playerSpeed": 1.5,
  "playerDecayRate": 0.002,
  "playerDecayCap": 0,
  "playerRecombineTime": 30,
  "playerMaxNickLength": 15,
  "playerDisconnectTime": -1,
  "splitVelocity": 780,

  // [MINIONS]
  // Custom minion settings
  // minionStartSize: Start size of minions (mass: val*val/100: 10 mass)
  // minionMaxStartSize: Maximum value of random start size for minions (set value higher than minionStartSize to enable)
  // minionCollideTeam: Determines whether minions colide with their team in the Teams gamemode (0: OFF, 1: ON)
  // serverMinions: Amount of minions each player gets once they spawn
  // defaultName: Default name for all minions if name is not specified using command (put <r> before the name for random skins!)
  // minionsOnLeaderboard: Whether or not to show minions on the leaderboard. (Set 0 to disable)
  "minionStartSize": 31.6227766017,
  "minionMaxStartSize": 31.6227766017,
  "minionCollideTeam": 0,
  "serverMinions": 0,
  "defaultName": "minion",
  "minionsOnLeaderboard": 0,

  // [Gamemode]
  // Custom gamemode settings
})
